File: pkg/transport/inmem/inmem.go
// Package inmem provides an in-process Transport driver. It is intended for
// unit tests, examples, and single-binary deployments. It is not safe for
// cross-process communication.
package inmem

import (
	"context"
	"errors"
	"sync"
	"sync/atomic"

	"github.com/google/uuid"

	"github.com/ModulationAI/agentflowbus/pkg/transport"
)

// Driver is an in-memory pub/sub broker.
type Driver struct {
	mu     sync.RWMutex
	subs   map[string][]*subscription
	rr     map[string]*uint64 // queue group round-robin counters, keyed by "subject\x00queue"
	closed bool
}

type subscription struct {
	driver  *Driver
	subject string
	queue   string
	handler transport.Handler
}

// New constructs an empty in-memory driver.
func New() *Driver {
	return &Driver{
		subs: make(map[string][]*subscription),
		rr:   make(map[string]*uint64),
	}
}

func (*Driver) Connect(context.Context) error { return nil }

func (d *Driver) Close() error {
	d.mu.Lock()
	d.closed = true
	d.subs = nil
	d.rr = nil
	d.mu.Unlock()
	return nil
}

func (*Driver) Capabilities() transport.Capabilities {
	return transport.Capabilities{
		Streaming:   true,
		Persistence: false,
		QueueGroup:  true,
		Headers:     true,
	}
}

func (d *Driver) Publish(ctx context.Context, msg *transport.RawMessage) error {
	if msg == nil {
		return errors.New("inmem: nil message")
	}

	d.mu.RLock()
	if d.closed {
		d.mu.RUnlock()
		return errors.New("inmem: driver closed")
	}
	subs := append([]*subscription(nil), d.subs[msg.Subject]...)
	d.mu.RUnlock()

	// Group by queue: empty queue == fan-out; non-empty queue == round-robin.
	var fanout []*subscription
	groups := map[string][]*subscription{}
	for _, s := range subs {
		if s.queue == "" {
			fanout = append(fanout, s)
		} else {
			groups[s.queue] = append(groups[s.queue], s)
		}
	}

	for _, s := range fanout {
		if err := s.handler(ctx, msg); err != nil {
			return err
		}
	}
	for q, members := range groups {
		idx := d.nextRR(msg.Subject, q, uint64(len(members)))
		if err := members[idx].handler(ctx, msg); err != nil {
			return err
		}
	}
	return nil
}

func (d *Driver) nextRR(subject, queue string, n uint64) uint64 {
	if n == 0 {
		return 0
	}
	key := subject + "\x00" + queue
	d.mu.Lock()
	c, ok := d.rr[key]
	if !ok {
		c = new(uint64)
		d.rr[key] = c
	}
	d.mu.Unlock()
	v := atomic.AddUint64(c, 1) - 1
	return v % n
}

func (d *Driver) Subscribe(_ context.Context, subject, queue string, h transport.Handler) (transport.Subscription, error) {
	if h == nil {
		return nil, errors.New("inmem: nil handler")
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.closed {
		return nil, errors.New("inmem: driver closed")
	}
	s := &subscription{driver: d, subject: subject, queue: queue, handler: h}
	d.subs[subject] = append(d.subs[subject], s)
	return s, nil
}

func (s *subscription) Unsubscribe() error {
	d := s.driver
	d.mu.Lock()
	defer d.mu.Unlock()
	list := d.subs[s.subject]
	for i, sub := range list {
		if sub == s {
			d.subs[s.subject] = append(list[:i], list[i+1:]...)
			return nil
		}
	}
	return nil
}

func (d *Driver) Request(ctx context.Context, msg *transport.RawMessage) (*transport.RawMessage, error) {
	// Reserved for v0.1 bus implementation; the bus layer wraps OpenInbox to
	// cover both request/reply and streaming.
	return nil, transport.ErrNotImplemented
}

func (d *Driver) OpenInbox(ctx context.Context) (transport.Inbox, error) {
	subject := "_INBOX.inmem." + uuid.NewString()
	box := &inbox{
		subject: subject,
		ch:      make(chan *transport.RawMessage, 64),
	}
	sub, err := d.Subscribe(ctx, subject, "", func(ctx context.Context, m *transport.RawMessage) error {
		select {
		case box.ch <- m:
			return nil
		case <-ctx.Done():
			return ctx.Err()
		}
	})
	if err != nil {
		return nil, err
	}
	box.sub = sub
	return box, nil
}

type inbox struct {
	subject string
	ch      chan *transport.RawMessage
	sub     transport.Subscription
	once    sync.Once
}

func (i *inbox) Subject() string { return i.subject }

func (i *inbox) Recv(ctx context.Context) (*transport.RawMessage, error) {
	select {
	case m, ok := <-i.ch:
		if !ok {
			return nil, errors.New("inmem inbox: closed")
		}
		return m, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

func (i *inbox) Close() error {
	var err error
	i.once.Do(func() {
		if i.sub != nil {
			err = i.sub.Unsubscribe()
		}
		close(i.ch)
	})
	return err
}

// Compile-time interface check.
var _ transport.Transport = (*Driver)(nil)

File: sdk/python/src/agentflowbus/transport/inmem.py
"""In-process pub/sub broker. Mirrors pkg/transport/inmem/inmem.go.

Intended for unit tests, examples, and single-binary deployments. Not safe for
cross-process communication.
"""
from __future__ import annotations

import asyncio
import uuid

from agentflowbus.transport.base import (
    Capabilities,
    Inbox,
    RawMessage,
    Subscription,
    Transport,
    TransportHandler,
)

# Module-level sentinel used to wake an inbox parked inside `recv` after `close`.
_CLOSED = object()


class InMemoryDriver:
    """Async in-memory pub/sub broker."""

    def __init__(self) -> None:
        self._lock = asyncio.Lock()
        self._subs: dict[str, list[_Subscription]] = {}
        self._rr: dict[str, int] = {}
        self._closed = False

    async def connect(self) -> None:
        return None

    async def close(self) -> None:
        async with self._lock:
            self._closed = True
            self._subs.clear()
            self._rr.clear()

    def capabilities(self) -> Capabilities:
        return Capabilities(
            streaming=True,
            persistence=False,
            queue_group=True,
            headers=True,
        )

    async def publish(self, msg: RawMessage) -> None:
        if msg is None:
            raise ValueError("inmem: nil message")
        async with self._lock:
            if self._closed:
                raise RuntimeError("inmem: driver closed")
            subs = list(self._subs.get(msg.subject, ()))

        # Fan-out for empty queue, round-robin per non-empty queue group.
        fanout: list[_Subscription] = []
        groups: dict[str, list[_Subscription]] = {}
        for s in subs:
            if not s.queue:
                fanout.append(s)
            else:
                groups.setdefault(s.queue, []).append(s)

        for s in fanout:
            await s.handler(msg)
        for q, members in groups.items():
            key = msg.subject + "\x00" + q
            counter = self._rr.get(key, 0)
            idx = counter % len(members)
            self._rr[key] = (counter + 1) & 0x3FFFFFFF
            await members[idx].handler(msg)

    async def subscribe(
        self,
        subject: str,
        queue: str,
        handler: TransportHandler,
    ) -> Subscription:
        if handler is None:
            raise ValueError("inmem: nil handler")
        async with self._lock:
            if self._closed:
                raise RuntimeError("inmem: driver closed")
            sub = _Subscription(self, subject, queue or "", handler)
            self._subs.setdefault(subject, []).append(sub)
            return sub

    async def request(
        self,
        msg: RawMessage,
        timeout: float | None = None,
    ) -> RawMessage:
        # Mirror Go: bus uses open_inbox for both request/reply and streaming.
        raise NotImplementedError("inmem: request not implemented; use bus.invoke")

    async def open_inbox(self) -> Inbox:
        subject = "_INBOX.inmem." + uuid.uuid4().hex
        box = _InMemoryInbox(self, subject)

        async def handler(m: RawMessage) -> None:
            await box._enqueue(m)

        box._sub = await self.subscribe(subject, "", handler)
        return box


class _Subscription:
    def __init__(
        self,
        driver: "InMemoryDriver",
        subject: str,
        queue: str,
        handler: TransportHandler,
    ) -> None:
        self.driver = driver
        self.subject = subject
        self.queue = queue
        self.handler = handler
        self._closed = False

    async def unsubscribe(self) -> None:
        if self._closed:
            return
        self._closed = True
        async with self.driver._lock:
            lst = self.driver._subs.get(self.subject, [])
            try:
                lst.remove(self)
            except ValueError:
                pass


class _InMemoryInbox:
    def __init__(self, driver: "InMemoryDriver", subject: str) -> None:
        self._driver = driver
        self._subject = subject
        self._queue: asyncio.Queue = asyncio.Queue()
        self._sub: Subscription | None = None
        self._closed = False

    @property
    def subject(self) -> str:
        return self._subject

    async def _enqueue(self, msg: RawMessage) -> None:
        if not self._closed:
            await self._queue.put(msg)

    async def recv(self, timeout: float | None = None) -> RawMessage:
        getter = self._queue.get()
        if timeout is not None:
            getter = asyncio.wait_for(getter, timeout)
        item = await getter
        if item is _CLOSED:
            await self._queue.put(_CLOSED)  # keep the queue poisoned for any future recv
            raise RuntimeError("inmem inbox: closed")
        return item

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self._sub is not None:
            try:
                await self._sub.unsubscribe()
            except Exception:
                pass
        await self._queue.put(_CLOSED)
